libr/util/r_pkcs7.c

```diff
@@ -0,0 +1,372 @@
+#include <stdlib.h>
+#include <string.h>
+#include <r_util.h>
+
+bool r_pkcs7_parse_certificaterevocationlists (RPKCS7CertificateRevocationLists *crls, RASN1Object *object) {
+	ut32 i;
+	if (!crls && !object) {
+		return false;
+	}
+	if (object->list.length > 0) {
+		crls->elements = (RX509CertificateRevocationList **) calloc (object->list.length, sizeof (RX509CertificateRevocationList*));
+		if (!crls->elements) {
+			return false;
+		}
+		crls->length = object->list.length;
+		for (i = 0; i < crls->length; ++i) {
+			crls->elements[i] = r_x509_parse_crl (object->list.objects[i]);
+		}
+	}
+	return true;
+}
+
+void r_pkcs7_free_certificaterevocationlists (RPKCS7CertificateRevocationLists *crls) {
+	ut32 i;
+	if (crls) {
+		for (i = 0; i < crls->length; ++i) {
+			r_x509_free_crl (crls->elements[i]);
+			crls->elements[i] = NULL;
+		}
+		free (crls->elements);
+		crls->elements = NULL;
+		// Used internally pkcs #7, so it should't free crls.
+	}
+}
+
+bool r_pkcs7_parse_extendedcertificatesandcertificates (RPKCS7ExtendedCertificatesAndCertificates *ecac, RASN1Object *object) {
+	ut32 i;
+	if (!ecac && !object) {
+		return false;
+	}
+	if (object->list.length > 0) {
+		ecac->elements = (RX509Certificate **) calloc (object->list.length, sizeof (RX509Certificate*));

```

Comment:
integer overflow. object->list.length I assume it comes from the file you are parsing. If this wrap to zero, afterwards the loop will use ecac->length and oob write will happen. Review the code to avoid this kind of issues please.